### ISSUE TITLE
Ref assemblies should not drop [Pure] attribute from APIs

### ISSUE THREAD

**Issue by AArnott** [dotnet/corefx#34778]
The Pure attribute is checked by Rule CA1806 in Microsoft.CodeQuality.Analyzers. The `[Pure]` attribute helps prevent many bugs, particular around APIs such as the immutable collections where folks frequently make such mistakes as this:

```cs
    public static void FOo()
    {
        var list = ImmutableList.Create<int>();
        list.Add(3);
    }
```

If `ImmutableList<T>.Add(T)` had the `[Pure]` attribute on it, the CA1806 analyzer mentioned above would produce an error in the above snippet, thereby helping the developer avoid a bug.

In fact the `Add` method *does* have a `[Pure]` attribute on it:

https://github.com/dotnet/corefx/blob/abc7e38a7a6bb0eed6a972127438685d7e9cbc98/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.cs#L222-L223

But the ref assembly drops these attributes, rendering all the effort we went to in tagging these pure methods to no effect.

In the below example, the call to `ImmutableList<int>.Add` does not produce a warning, but the next call to the sample's own `Add` method _does_ produce a warning, because the `[Pure]` attribute is actually there.

```cs
    public static void FOo()
    {
        var list = ImmutableList.Create<int>();
        list.Add(3);
        Add(3, 5);
    }

    [Pure]
    public static int Add(int a, int b) => a + b;
```

Can whatever build process that generates these ref assemblies be fixed to preserve this attribute?

_Originally posted by @AArnott in https://github.com/dotnet/corefx/pull/31071_

**Comment by tannergooding** [comment]
Do the reference assemblies produced by the C# compiler include these types of attributes. CC. @jaredpar?

**Comment by AArnott** [comment]
@tannergooding I just confirmed that csc *does* emit `[Pure]` attributes in ref assemblies. 

**But** the `PureAttribute` class itself has `[Conditional("CONTRACTS_FULL")]` applied, so unless that constant is defined, the attribute will be dropped from the ref and the impl assembly alike.

**Comment by Grauenwolf** [comment]
Can we get that changed? I don't see any reason to strip away the PureAttribute without warning the developer that they need to add a hidden compiler flag. That is just going to lead to confusion.

**Comment by jaredpar** [comment]
> Do the reference assemblies produced by the C# compiler include these types of attributes.

Yes we do keep (as @AArnott confirmed). Consider that if we did not then it would significantly hamper analyzers. Any analyzer based on attributes would essentially be un-usable in the face of reference assemblies.

**Comment by tannergooding** [comment]
Given that the C# ref assembly feature does preserve the attributes (where applicable), this should probably be moved to dotnet/arcade so we can update `GenFacades`.

I think the removal of `[Conditional("CONTRACTS_FULL")]` from the `PureAttribute` class is likely a separate proposal (which would likely live in CoreFX).

CC. @ericstj, @ViktorHofer

**Comment by AArnott** [comment]
Is it `GenFacades` that needs to define the symbol or corefx? Are the assemblies generated from the corefx repo, and if so, wouldn't corefx be the repo that needs to be changed to define the symbol?

Or we could resolve this issue by removing the `[Conditional]` attribute. Either way, it seems like corefx is the repo to change, isn't it?

**Comment by tannergooding** [comment]
The ref assemblies in CoreFX are special and are not generated by the C# compiler, instead they are (outside a couple exceptions) generated by some special tooling in the arcade repo. The root issue here is that the tooling in arcade is doing something different than the C# compiler (specifically, it isn't counting attributes as part of the public API). There have been other cases of this in the past, and we have had to fix those cases as well (such as counting private fields of structs as part of this public API).

So, ultimately we likely need two (possibly three issues):
* An issue in dotnet/arcade tracking us updating the tooling that generates the CoreFX reference assemblies (ensuring that generated reference assemblies include attributes as part of the public API)
* An issue in dotnet/corefx tracking us consuming the new tooling and regenerating all of our reference assemblies (ensuring that the shipping reference assemblies now include attributes)
* Potentially an issue in dotnet/corefx tracking the removal of `[Conditional("CONTRACTS_FULL")]` from `PureAttribute` (which would ensure that this attribute ends up in the ref assembly)

**Comment by ericstj** [comment]
> The ref assemblies in CoreFX are special and are not generated by the C# compiler, instead they are (outside a couple exceptions) generated by some special tooling in the arcade repo.

That is not true.  We use tools to generate source code and that source code is checked in and directly compiled by CSC.

Did you mean GenAPI?  GenAPI will persist the pure attribute if it exists.  That's controlled by an exclusion list and PureAttribute isn't in it.  https://github.com/dotnet/corefx/blob/master/eng/DefaultGenApiDocIds.txt

If the PureAttribute is present in the source you can run GenAPI and it will add it to the checked in source file.  If the PureAttribute is in the checked in source file it will be preserved in the compiled reference assembly.

**Comment by AArnott** [comment]
> If the PureAttribute is present in the source you can run GenAPI and it will add it to the checked in source file. If the PureAttribute is in the checked in source file it will be preserved in the compiled reference assembly.

I assume you mean after the `[Conditional]` attribute is removed, or if the `csc` tool is invoked to generate that ref assembly with the `CONTRACTS_FULL` symbol defined, right @ericsj?

**Comment by ericstj** [comment]
This worked for me:
1. Add CONTRACTS_FULL to  https://github.com/dotnet/corefx/blob/8fea941db55fd79635fde1a54058620ddac16961/src/System.Collections.Immutable/src/System.Collections.Immutable.csproj#L12
2. dotnet msbuild
3. dotnet msbuild /t:GenerateReferenceSource
4. observe the differences to the generated source file in the rfeference assembly folder (and add back the ifdefs as you wish)
5. add a dependency to the reference assembly on System.Diagnostics.Contracts so that it builds again.  Add the CONTRACTS_FULL define as well.
6. Build the reference assembly.
7.  Observe pureattributes in your reference assembly.

**Comment by ViktorHofer** [comment]
@safern @ericstj do we want to fix this for 3.0? sounds not irrelevant.
